api/src/ipc.c

```diff
@@ -108,6 +108,13 @@ static void ipc_free_allocated_completed_tokens(uint32_t tokens)
 
 static int ipc_waitfor(int (*cond)(uint32_t have, uint32_t expect), uint32_t wait_tokens, uint32_t * done_tokens, uint32_t timeout_ms)
 {
+    uvisor_spin_lock(ipc_tokens_lock());
+    /* Check we are not waiting for some unallocated tokens */
+    if ((wait_tokens & *ipc_allocated_tokens()) != wait_tokens) {
+        uvisor_spin_unlock(ipc_tokens_lock());
+        return UVISOR_ERROR_INVALID_PARAMETERS;
+    }
+    uvisor_spin_unlock(ipc_tokens_lock());
```

Comment:
A token might get freed or allocated between here and when we read the completed tokens. This block of code doesn't prevent waiting on tokens being allocated in all cases due to this race condition. The way to avoid this would be to hold the lock the entire time one is spinning, but that would prevent any change to the completed tokens and make the waitfor useless.